### ISSUE TITLE
Socket errors on Ubuntu host

### ISSUE THREAD

**Issue by zblesk** [dotnet/aspnetcore#30787]
### Description

**tl;dr** I have a Vue,JS SPA built on .NET 5/ASP.NET Core 5. 
When developing and running on a Windows host, everything works normally; when running on a Ubuntu Server 18.04 LTS host, socket exceptions pop up. 

**Background:** I have tried various configurations and combinations of hosts, and this error pops up only on Ubuntu Server hosts, not on Windows. On both platforms, I tried both running the app in dev mode (with `dotnet run`) and publishing first, _then_ running (with `dotnet <mydll>`). When publishing this way, I just used `dotnet publish` (i.e. without specifying any target, like `linux-x64`). 

Besides that, I've also tried publishing manually with `-r linux-x64 --self-contained false`. I've also built docker images, using `mcr.microsoft.com/dotnet/runtime:5.0-focal-amd64` and `mcr.microsoft.com/dotnet/runtime-deps:5.0-focal-amd64` as target runtime images (based on [the aspnet dockerfile samples from here](https://github.com/dotnet/dotnet-docker/tree/main/samples/aspnetapp)).

All of these configurations, I've tried both behind a nginx proxy, and also directly running on port 80. 
(I tried them basically in the reverse order I'm mentioning here: since I had no problems runing .NET Core 2 and 3 apps on the same system and behind nginx, that was the first thing I tried, not expecting any issues; when issues kept popping up, I simpified down to a build made directly on the server, running directly on port 80, to exclude as many possible issues as I could.)

In all of those scenarios, the errors occurred. I haven't found any underlying connection - just clicking around the SPA site,  issuing requests to the API without modifying any data, sometimes the requests went through without any issue, and quite often they failed. At times, approx. 80%-90% of the requests failed (only on Ubuntu Server host.) 
There were no other web servers running on the machine (which is a VM in Azure) at the time, and neither was it under any significant load.

**The logs and `dotnet info` outputs quoted here** are from the simlpest scenario I tried: a `dotnet publish` directly on the server, with a subsequent `sudo dotnet <mydll>`. (Using `sudo` only so I could bind to port 80.) 
I've also turned the `Microsoft.AspNetCore` log level to `Debug`.

Any help is appreciated.

### Exceptions

Two messages are popping up most often: 

**Connection processing ended abnormally. System.InvalidOperationException: Handle is already used by another Socket.** 

```
2021-03-09 20:12:41.176 +00:00 [Information] HTTP "GET" "/api/users" responded 200 in 13.7170 ms
2021-03-09 20:12:41.177 +00:00 [Debug] Execution plan of "authorization" filters (in the following order): ["None"]
2021-03-09 20:12:41.177 +00:00 [Debug] Connection id ""0HM736PP9IQ0N"" completed keep alive response.
2021-03-09 20:12:41.177 +00:00 [Debug] Execution plan of "resource" filters (in the following order): ["None"]
2021-03-09 20:12:41.179 +00:00 [Information] Request finished HTTP/1.1 GET http://bct.zble.sk/api/users - - - 200 - application/json;+charset=utf-8 18.7564ms
2021-03-09 20:12:41.179 +00:00 [Debug] Execution plan of "action" filters (in the following order): ["Microsoft.AspNetCore.Mvc.Filters.ControllerActionFilter (Order: -2147483648)", "Microsoft.AspNetCore.Mvc.ModelBinding.UnsupportedContentTypeFilter (Order: -3000)"]
2021-03-09 20:12:41.180 +00:00 [Debug] Execution plan of "exception" filters (in the following order): ["None"]
2021-03-09 20:12:41.180 +00:00 [Debug] Execution plan of "result" filters (in the following order): ["None"]
2021-03-09 20:12:41.180 +00:00 [Debug] Executing controller factory for controller "Obskurnee.Controllers.AdminController" ("Obskurnee")
2021-03-09 20:12:41.180 +00:00 [Debug] Executed controller factory for controller "Obskurnee.Controllers.AdminController" ("Obskurnee")
2021-03-09 20:12:41.182 +00:00 [Information] Executing JsonResult, writing value of type '"<>f__AnonymousType1`2[[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Object, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]"'.
2021-03-09 20:12:41.183 +00:00 [Information] Executed action "Obskurnee.Controllers.AdminController.GetAdminInfo (Obskurnee)" in 3.0293ms
2021-03-09 20:12:41.184 +00:00 [Information] Executed endpoint '"Obskurnee.Controllers.AdminController.GetAdminInfo (Obskurnee)"'
2021-03-09 20:12:41.185 +00:00 [Information] HTTP "GET" "/api/admin" responded 200 in 11.9339 ms
2021-03-09 20:12:41.186 +00:00 [Debug] Connection id ""0HM736PP9IQ18"" completed keep alive response.
2021-03-09 20:12:41.187 +00:00 [Information] Request finished HTTP/1.1 GET http://bct.zble.sk/api/admin - - - 200 - application/json;+charset=utf-8 18.2253ms
2021-03-09 20:12:41.211 +00:00 [Debug] Connection id ""0HM736PP9IQ19"" communication error.
System.InvalidOperationException: Handle is already used by another Socket.
   at System.Net.Sockets.SocketAsyncEngine.RegisterCore(IntPtr socketHandle, SocketAsyncContext context)
   at System.Net.Sockets.SocketAsyncEngine.RegisterSocket(IntPtr socketHandle, SocketAsyncContext context)
   at System.Net.Sockets.SocketAsyncContext.Register()
   at System.Net.Sockets.SocketAsyncContext.OperationQueue`1.StartAsyncOperation(SocketAsyncContext context, TOperation operation, Int32 observedSequenceNumber, CancellationToken cancellationToken)
   at System.Net.Sockets.SocketAsyncContext.ReceiveAsync(Memory`1 buffer, SocketFlags flags, Int32& bytesReceived, Action`5 callback, CancellationToken cancellationToken)
   at System.Net.Sockets.SocketAsyncEventArgs.DoOperationReceive(SafeSocketHandle handle, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.ReceiveAsync(SocketAsyncEventArgs e, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketReceiver.WaitForDataAsync()
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection.ProcessReceives()
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection.DoReceive()
2021-03-09 20:12:41.215 +00:00 [Debug] Connection id ""0HM736PP9IQ19"" accepted.
2021-03-09 20:12:41.217 +00:00 [Debug] Connection id ""0HM736PP9IQ19"" started.
2021-03-09 20:12:41.217 +00:00 [Debug] Connection id ""0HM736PP9IQ19"" sending FIN because: ""The client closed the connection.""
2021-03-09 20:12:41.217 +00:00 [Debug] Connection id ""0HM736PP9IQ19"" disconnecting.
2021-03-09 20:12:41.218 +00:00 [Warning] Connection processing ended abnormally.
System.InvalidOperationException: Handle is already used by another Socket.
   at System.Net.Sockets.SocketAsyncEngine.RegisterCore(IntPtr socketHandle, SocketAsyncContext context)
   at System.Net.Sockets.SocketAsyncEngine.RegisterSocket(IntPtr socketHandle, SocketAsyncContext context)
   at System.Net.Sockets.SocketAsyncContext.Register()
   at System.Net.Sockets.SocketAsyncContext.OperationQueue`1.StartAsyncOperation(SocketAsyncContext context, TOperation operation, Int32 observedSequenceNumber, CancellationToken cancellationToken)
   at System.Net.Sockets.SocketAsyncContext.ReceiveAsync(Memory`1 buffer, SocketFlags flags, Int32& bytesReceived, Action`5 callback, CancellationToken cancellationToken)
   at System.Net.Sockets.SocketAsyncEventArgs.DoOperationReceive(SafeSocketHandle handle, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.ReceiveAsync(SocketAsyncEventArgs e, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketReceiver.WaitForDataAsync()
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection.ProcessReceives()
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection.DoReceive()
   at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
   at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
   at System.IO.Pipelines.Pipe.ReadAsync(CancellationToken token)
   at System.IO.Pipelines.Pipe.DefaultPipeReader.ReadAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.BeginRead(ValueTask`1& awaitable)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequestsAsync[TContext](IHttpApplication`1 application)
2021-03-09 20:12:41.219 +00:00 [Debug] Connection id ""0HM736PP9IQ19"" stopped.

```

The other exception is **Connection processing ended abnormally. System.Net.Sockets.SocketException (125): Operation canceled**

```
2021-03-09 19:50:13.181 +00:00 [Information] HTTP "GET" "/api/accounts/context" responded 200 in 14.0045 ms
2021-03-09 19:50:13.183 +00:00 [Debug] Connection id ""0HM733JHLKC9T"" completed keep alive response.
2021-03-09 19:50:13.184 +00:00 [Information] Request finished HTTP/1.1 GET http://bct.zble.sk/api/accounts/context - - - 200 - application/json;+charset=utf-8 31.9169ms
2021-03-09 19:50:13.186 +00:00 [Debug] Connection id ""0HM733JHLKC9T"" sending FIN because: ""Operation canceled""
2021-03-09 19:50:13.186 +00:00 [Information] Executed action "Obskurnee.Controllers.HomeController.LandingPage (Obskurnee)" in 25.802ms
2021-03-09 19:50:13.188 +00:00 [Information] Executed endpoint '"Obskurnee.Controllers.HomeController.LandingPage (Obskurnee)"'
2021-03-09 19:50:13.189 +00:00 [Information] HTTP "GET" "/api/home" responded 200 in 31.0896 ms
2021-03-09 19:50:13.190 +00:00 [Warning] Connection processing ended abnormally.
System.Net.Sockets.SocketException (125): Operation canceled
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketAwaitableEventArgs.<GetResult>g__ThrowSocketException|7_0(SocketError e)
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketAwaitableEventArgs.GetResult()
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection.ProcessSends()
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection.DoSend()
   at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
   at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
   at System.IO.Pipelines.Pipe.GetReadAsyncResult()
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequestsAsync[TContext](IHttpApplication`1 application)
2021-03-09 19:50:13.191 +00:00 [Debug] Connection id ""0HM733JHLKC9T"" disconnecting.
2021-03-09 19:50:13.192 +00:00 [Debug] Connection id ""0HM733JHLKC9T"" stopped.
2021-03-09 19:50:13.192 +00:00 [Debug] Connection id ""0HM733JHLKC9U"" completed keep alive response.
2021-03-09 19:50:13.193 +00:00 [Debug] Connection id ""0HM733JHLKC9U"" received FIN.
2021-03-09 19:50:13.193 +00:00 [Debug] Connection id ""0HM733JHLKC9U"" sending FIN because: ""The client closed the connection.""
2021-03-09 19:50:13.193 +00:00 [Debug] Connection id ""0HM733JHLKC9U"" disconnecting.
2021-03-09 19:50:13.194 +00:00 [Information] Request finished HTTP/1.1 GET http://bct.zble.sk/api/home - - - 200 - application/json;+charset=utf-8 38.8609ms
2021-03-09 19:50:13.195 +00:00 [Debug] Connection id ""0HM733JHLKC9U"" stopped.
```

### Further technical details

Output of `dotnet --info` on the host the logs come from: 

```
~> dotnet --info
.NET SDK (reflecting any global.json):
 Version:   5.0.200
 Commit:    70b3e65d53

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  18.04
 OS Platform: Linux
 RID:         ubuntu.18.04-x64
 Base Path:   /usr/share/dotnet/sdk/5.0.200/

Host (useful for support):
  Version: 5.0.3
  Commit:  eae88cc11b

.NET SDKs installed:
  5.0.200 [/usr/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 5.0.3 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 5.0.3 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download
```

**Comment by BrennanConroy** [comment]
Can you share your repro app?

**Comment by zblesk** [comment]
Unfortunately, I haven't been able to create a good repro app - the issue doesn't seem to occur in a trivial one.  
I understand if you can't do anything about it in this case, and will close this issue. 
Could you maybe give me a  few pointers on how could I go about debugging this issue myself? (For example, some key words I should look up, regarding diagnosing issues of a published app on an Ubuntu server. I do have experience troubleshooting things on a Windows dev box, but up until this point, I just published to Ubuntu Server and it just ran without problems, so I'm not sure where to start.)

**Comment by BrennanConroy** [comment]
Are you creating a Socket anywhere in your code? Are you doing any systemd activation in your app?

VS should be able to attach to a Linux process using remote debugging.

**Comment by zblesk** [comment]
I'm doing neither of those. 
Thank you, I will try the remote debugging, then.

**Comment by karelz** [comment]
FYI: https://github.com/dotnet/runtime/issues/56750#issuecomment-930896613 and https://github.com/dotnet/runtime/issues/56750#issuecomment-931163514 explain the narrowed down problem and workaround. The real root cause was not identified.
